### PR TITLE
fix(supervisor): stop trusting a stale supervisor pid after reboot or pid reuse

### PR DESCRIPTION
**Pull request by jdx**
https://entire.io/gh/jdx/pitchfork/trails/3

## Problem

The supervisor's entry in `state.toml` survives crashes and reboots. Every command that acted on it trusted the bare pid: `kill(pid, 0)` succeeding meant "the supervisor is running". After a reboot, low pids are handed to early system daemons, so a user in [discussion #877](https://github.com/jdx/pitchfork/discussions/877) saw:

- every CLI command failing with `Failed to connect to supervisor: failed to connect to supervisor after 5 attempts`, because auto-start believed a supervisor was already up
- `pitchfork supervisor start` and `run` refusing with `Pitchfork supervisor is already running with pid 1207. Use --force to replace it.`
- `pitchfork supervisor stop` printing `Stopped pitchfork daemon with pid 1207` after sending SIGTERM to `AMPDeviceDiscoveryAgent`, an unrelated macOS LaunchAgent that now owned that pid

The same thing happens without a reboot whenever the supervisor dies and the OS reuses its pid.

## Fix

The supervisor already records its `start_time` (kernel start token) and `boot_time` alongside its pid. This change makes every reader check them, using the same start-token identity rule the daemon paths have used since #632.

- `supervisor stop` on a stale record now clears the entry and reports `Pitchfork daemon with pid N is not running (cleaned up stale state)`. Nothing is signalled.
- `supervisor start` and `supervisor run` replace a stale record without `--force`.
- Ordinary commands auto-start the supervisor past a stale record instead of failing to connect.
- `supervisor status --json` reports `down` rather than `unknown` for a stale record.
- When the record is genuine and `--force` or `stop` is used, the kill is bound to the recorded process generation. On Linux the identity check binds a pidfd and every signal, including the SIGKILL escalation, is sent through it, so a pid recycled after the check gets `ESRCH` instead of the signal. Windows holds a process handle for the same effect. Other Unix platforms have no pidfd, so they re-validate the start token immediately before the first signal and again before the SIGKILL escalation, the tightest window those platforms allow.
- If the supervisor is alive but its start token cannot be read, `stop`, `start`, and `run` report an error and leave the record in place rather than treating it as dead, so a running supervisor is never orphaned by a failed identity read.

Identity rules, in order:

1. If both the recorded and the live start token are readable, they alone decide. Boot time is not allowed to veto matching tokens, because on Linux and macOS the reported boot time is derived from the realtime clock and moves under NTP steps or sleep/resume while the supervisor keeps running.
2. If a token is missing on either side, a recorded boot time from a different boot marks the record stale.
3. A record written by pitchfork older than v2.18.0 has neither field. It is trusted only if the live process is a pitchfork binary, and the kill is then bound to the start token observed at that moment. There is no longer any path that signals a bare pid.

## Validation

- Unit tests cover the identity rule: same generation, recycled pid within a boot and across a reboot, clock steps with matching tokens, boot-time fallback when a token is missing, and the legacy name check (including the `AMPDeviceDiscoveryAgent` case).
- Bats tests in `test/supervisor.bats` point a stale supervisor record at a bystander `sleep` process and assert that `stop`, `start`, and auto-start all leave it alive and recover, for records with a wrong start token, a previous boot time, and no identity at all. They fail on the unpatched binary and pass with the fix.
- Full run on Linux: all Rust tests and all bats tests pass, clippy with `-D warnings` and `cargo fmt --check` are clean, `mise run render` produced no changes. Not exercised on macOS or Windows, though the macOS and Windows start-token readers already exist and are used by the daemon paths.

Note: `mise run ci-dev` could not complete locally because `hk fix --all` (hk 2.0.1) fails to evaluate `hk.pkl` on an untouched tree; the individual checks above were run instead.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-fable-5-1; version: unavailable.*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved supervisor process tracking to prevent signaling unrelated processes after PID reuse or stale state.
  * Supervisor status now more accurately distinguishes running, stopped, unknown, and stale records.
  * Stale supervisor records are cleaned up safely, allowing starts and regular commands to recover automatically.
  * Legacy records are validated using process identity and title checks before being trusted.

---

> [!NOTE]
> **Medium Risk**
> Touches supervisor lifecycle and cross-platform process signalling with new identity gates; behavior is heavily tested but mistakes could still affect stop/start or signal the wrong PID.
> 
> **Overview**
> Fixes cases where a **surviving `state.toml` supervisor entry** pointed at an unrelated live PID (reboot or reuse), which blocked auto-start, refused start/run without `--force`, or **sent stop signals to the wrong process**.
> 
> Supervisor actions now treat the state record as **live only when process identity matches** (`start_time` / `boot_time`, with a weaker legacy name check for old records). Stale entries are cleared without signalling; start, auto-start, and status follow that model instead of trusting `kill(pid, 0)`.
> 
> When a real supervisor must be stopped or replaced, kills go through **`kill_if_start_time_matches`** (pidfd-pinned signals on Linux, handle pinning on Windows, start-token re-checks before SIGKILL elsewhere). Unreadable identity on a still-running PID **errors and keeps the record**, avoiding orphaning a running supervisor. Bats and unit tests cover stale-PID bystanders and identity edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 522850b157d485515d50326795b02796af7afedb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/procs.rs [hunk 1]
```diff
@@ -221,6 +221,81 @@ impl Procs {
...
+        match self.start_time(pid) {
+            Some(current) if current == expected_start_time => {}
+            current => {
+                warn!(
+                    "pid {pid} is not the recorded process (start time {current:?}, expected {expected_start_time}); not signalling it"
+                );
+                return Ok(false);
+            }
+        }
+        self.kill(pid, stop_signal, stop_timeout)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/procs.rs]: Agreed: a pidfd keeps the struct pid alive but the number is freed on reap, so it did not protect the numeric kill. In 9356985 the Linux path now sends every signal (stop signal and the SIGKILL escalation) through the pidfd bound by the identity check and polls the pidfd for exit, so a recycled pid gets ESRCH instead of a signal.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/cli/supervisor/mod.rs [hunk 2]
```diff
@@ -47,42 +49,61 @@ impl Supervisor {
...
-            }
-        } else {
-            Ok(KillOrStopOutcome::StillRunning)
         }
-    } else {
-        Ok(KillOrStopOutcome::AlreadyDead)
+        // A record written by a supervisor that predates start-time tracking
+        // has nothing to bind to; it was already accepted as live above (same
+        // boot, if known), so stop it the way it always was.
+        None => PROCS.kill_async(existing_pid, stop_signal, None).await,
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/supervisor/mod.rs]: Fixed in 9356985: the unbound `kill_async` fallback is gone. A legacy record with neither field is trusted only when the live process is a pitchfork binary, and the kill is then bound to the start token observed at that moment via `kill_if_start_time_matches_async`. No path signals a bare pid any more.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/mod.rs [hunk 3]
```diff
@@ -160,14 +160,74 @@ pub static SUPERVISOR: Lazy<Supervisor> =
...
+pub(crate) fn supervisor_identity_matches(
+    recorded_start_time: Option<u64>,
+    current_start_time: Option<u64>,
+    recorded_boot_time: Option<u64>,
+    current_boot_time: u64,
+) -> bool {
+    let same_boot = recorded_boot_time
+        .is_none_or(|recorded| recorded.abs_diff(current_boot_time) <= BOOT_TIME_TOLERANCE_SECS);
+    same_boot && signalling_pid_is_authorized(recorded_start_time, current_start_time)
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/mod.rs:257]: Good catch. In 9356985 start tokens alone decide identity when both are readable, matching the existing `process_identity_matches` rule; boot time is only consulted as a fallback when a token is missing on either side, so a clock step after sleep/resume or NTP can no longer mark a live supervisor stale. Unit tests cover the clock-step case.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/cli/supervisor/mod.rs [hunk 4]
```diff
@@ -47,42 +49,61 @@ impl Supervisor {
...
+    };
+    match killed {
+        Ok(true) => Ok(KillOrStopOutcome::Killed),
+        Ok(false) => Ok(KillOrStopOutcome::AlreadyDead),
+        Err(e) => Err(miette::miette!("{e}. Try rerun with sudo.")),
     }
 }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/supervisor/mod.rs:103]: Fixed in 9356985: `verify_start_time_before_signal` now distinguishes three outcomes. A matching token signals; a different token or a gone process is `Ok(false)` (nothing of ours left); an unreadable token on a live process is an `Err`, which `stop`/`start`/`run` propagate, so the state record is kept and no second supervisor is started.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/procs.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdx) [reply src/procs.rs]: Fixed in 9356985. Linux fails closed when a pidfd cannot be opened (error, not a numeric fallback) and signals only through the pidfd. On other Unix platforms `kill` now takes the expected start token and re-validates it immediately before the SIGKILL escalation, so a pid recycled during the stop timeout is left alone.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/cli/supervisor/mod.rs [hunk 5]
```diff
@@ -47,42 +49,57 @@ impl Supervisor {
...
 
 pub async fn resolve_existing_supervisor(force: bool) -> Result<(Option<u32>, KillOrStopOutcome)> {
-    let existing_pid = existing_supervisor_pid()?;
-    let outcome = if let Some(pid) = existing_pid {
-        kill_or_stop(pid, force).await?
+    let record = existing_supervisor()?;
+    let existing_pid = record.as_ref().and_then(|d| d.pid);
+    let outcome = if let Some(record) = &record {
+        kill_or_stop(record, force).await?
     } else {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/supervisor/mod.rs:86]: Fixed in 7cd1926: when a legacy record has no recorded start time and the live token cannot be read, `kill_or_stop` now returns an error instead of passing `None` to the kill helper. `stop`, `start --force` and `run --force` propagate it, so the record is kept and no second supervisor is started, consistent with `verify_start_time_before_signal`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
